pandas/io/formats/xml.py

```diff
@@ -0,0 +1,763 @@
+        )
+
+        if self.stylesheet:
+            self.out_xml = self.transform_doc()
+
+        return self.out_xml
+
+    def validate_columns(self) -> None:
+        """
+        Validate elems_cols and attrs_cols.
+
+        This method will check if columns is list-like.
+
+        Raises
+        ------
+        ValueError
+            * If value is not a list and less then length of nodes.
+        """
+        if self.attr_cols and not is_list_like(self.attr_cols):
+            raise TypeError(
+                f"{type(self.attr_cols).__name__} is not a valid type for attr_cols"
+            )
+
+        if self.elem_cols and not is_list_like(self.elem_cols):
+            raise TypeError(
+                f"{type(self.elem_cols).__name__} is not a valid type for elem_cols"
+            )
+
+    def validate_encoding(self) -> None:
+        """
+        Validate encoding.
+
+        This method will check if encoding is among listed under codecs.
+
+        Raises
+        ------
+        LookupError
+            * If encoding is not available in codecs.
+        """
+
+        try:
+            codecs.lookup(self.encoding)
+        except LookupError as e:
+            raise e
+
+    def process_dataframe(self) -> dict:
+        """
+        Adjust Data Frame to fit xml output.
+
+        This method will adjust underlying data frame for xml output,
+        including replacing missing entities and including indexes.
+        """
+
+        na_dict = {"None": self.na_rep, "NaN": self.na_rep, "nan": self.na_rep}
+
+        df = (
+            (self.fmt.frame.reset_index().applymap(str).replace(na_dict))
+            if self.index
+            else self.fmt.frame.applymap(str).replace(na_dict)
+        )
+
+        return df.to_dict(orient="index")
+
+    def convert_empty_str_key(self) -> None:
+        """
+        Replace zero-lengh string in `namespaces`.
+
+        This method will replce '' with None to align to `lxml`
+        requirement that empty string prefixes are not allowed.
+        """
+
+        if self.namespaces and "" in self.namespaces.keys():
+            self.namespaces[None] = self.namespaces.pop("", "default")
+
+    def handle_indexes(self) -> None:
+        """
+        Handle indexes.
+
+        This method will add indexes into attr_cols or elem_cols.
+        """
+        indexes = [x for x in self.frame_dicts[0].keys() if x not in self.orig_cols]
+
+        if self.attr_cols and self.index:
+            self.attr_cols = list(indexes) + self.attr_cols
+
+        if self.elem_cols and self.index:
+            self.elem_cols = list(indexes) + self.elem_cols
+
+    def get_prefix_uri(self) -> str:
+        """
+        Get uri of namespace prefix.
+
+        This method retrieves corresponding URI to prefix in namespaces.
+
+        Raises
+        ------
+        ValueError
+            *If prefix is not included in namespace dict.
+        """
+
+        uri = ""
```

Comment:
seems like a lot of these methods are duplicated can you make a base class to avoid this.

Reply:
Will look into over next day or two.